### PR TITLE
chore: add -Wthread-analysis and annotate (part 1/2)

### LINKED ISSUE #3448: enable -Wthread-safety for clang
Currently all the flows that use  `ABSL_GUARD_XXX` variants do not benefit from the flag because we don't use `-Wthread-safety` on `clang builds`.

Fix that, and if there are any fishy flows fix them as well.

### PR DESCRIPTION
**Pull request by kostasrim**
In order for our mutex classes to work with [clang's ThreadSafety analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#thread-safety-analysis) (used by ABSL_GUARD_XXX functions) we need to properly annotate them and enable the check. This PR adds the missing annotations.

Part 1/2. Partially fixes #3448

* enable -Wthread-analysis
* add missing annotations
* small fixes

There are *no* functional changes.

### CODE REVIEW

File: src/server/cluster/incoming_slot_migration.h [hunk 1]
```diff
@@ -47,12 +47,12 @@ class IncomingSlotMigration {
     return source_id_;
   }
 
-  void ReportError(dfly::GenericError err) {
+  void ReportError(dfly::GenericError err) ABSL_LOCKS_EXCLUDED(error_mu_) {
```
Comment (kostasrim) [inline src/server/cluster/incoming_slot_migration.h:50]: //TODO test negative capabilities via `!error_mu_` for stronger invariants. See https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#negative-capabilities

Comment (kostasrim) [reply src/server/cluster/incoming_slot_migration.h:50]: It's an experimental flag so I rather not introduce it until it's stable. IMO it's far better though in the static analysis.

File: src/server/cluster/cluster_family.h [hunk 2]
```diff
@@ -62,7 +62,8 @@ class ClusterFamily {
   void DflyClusterFlushSlots(CmdArgList args, ConnectionContext* cntx);
 
  private:  // Slots migration section
-  void DflySlotMigrationStatus(CmdArgList args, ConnectionContext* cntx);
```
Comment (kostasrim) [inline src/server/cluster/cluster_family.h:65]: There are a lot of warnings + actually cases where we forget to lock. I will fix them in a separate PR this is already getting big

File: src/server/server_family.cc [hunk 3]
```diff
@@ -1500,23 +1501,21 @@ GenericError ServerFamily::DoSaveCheckAndStart(bool new_version, string_view bas
 }
 
 GenericError ServerFamily::WaitUntilSaveFinished(Transaction* trans, bool ignore_state) {
+  util::fb2::LockGuard lk(save_mu_);
   save_controller_->WaitAllSnapshots();
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/server_family.cc:1505]: This was *not* a bug but rather a weird way to allow polling with `INFO` while we are waiting for the snapshots to finish. :face_exhaling:

**Review by dranikpg — APPROVED** [review]
I was suprised myself that absl lockables expect you to use custom holder/lock classes instead of std guards 😕

**Comment by kostasrim** [comment]
> I was suprised myself that absl lockables expect you to use custom holder/lock classes instead of std guards 😕

it's because we don't use libc++

**Review by romange — APPROVED** [review]

**Pull request merged**